llvm/include/llvm/CodeGen/MachineBasicBlock.h

```diff
@@ -673,10 +677,17 @@ class MachineBasicBlock
   /// Returns true if this block ends any section.
   bool isEndSection() const { return IsEndSection; }
 
+  /// Returns true if this block contains the first instruction of its section.
+  bool isFirstNonEmptyBBInSection() const { return IsFirstNonEmptyBBInSection; }
```

Comment:
I'm hesitant to add another maintained value to `MachineBasicBlock` because it introduces one more point of failure for silently invalidated invariants if subsequent MF passes don't properly update the value (for example, [`BranchRelaxation`](https://github.com/llvm/llvm-project/blob/8b6764fdc0c9550e3d8033006a4acfb466f74840/llvm/lib/CodeGen/BranchRelaxation.cpp#L267-L269) inserts basic blocks, potentially invalidating `isBeginSection`).

It looks like this value is only needed in `BasicBlockSections`; would it make sense to compute it directly there?

Reply:
That's fair. The intuition was that we track very similar properties already there, and that we get this value for free when iterating over the blocks anyways. But less state is good, so if you prefer computing it locally, I will move it.